GH-33953: [Java] Pass custom headers on every request

aiguofer opened the pull request:
<!--
Thanks for opening a pull request!
If this is your first pull request you can find detailed information on how 
to contribute here:
  * [New Contributor's Guide](https://arrow.apache.org/docs/dev/developers/guide/step_by_step/pr_lifecycle.html#reviews-and-merge-of-the-pull-request)
  * [Contributing Overview](https://arrow.apache.org/docs/dev/developers/overview.html)


If this is not a [minor PR](https://github.com/apache/arrow/blob/master/CONTRIBUTING.md#Minor-Fixes). Could you open an issue for this pull request on GitHub? https://github.com/apache/arrow/issues/new/choose

Opening GitHub issues ahead of time contributes to the [Openness](http://theapacheway.com/open/#:~:text=Openness%20allows%20new%20users%20the,must%20happen%20in%20the%20open.) of the Apache Arrow project.

Then could you also rename the pull request title in the following format?

    GH-${GITHUB_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

or

    MINOR: [${COMPONENT}] ${SUMMARY}

In the case of PARQUET issues on JIRA the title also supports:

    PARQUET-${JIRA_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

-->

### Rationale for this change

Some flight requests don't send custom headers. This PR should fix that.

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

### What changes are included in this PR?

Ensure custom headers are sent across on every request.

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

### Are these changes tested?

No

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

### Are there any user-facing changes?

Custom headers should now be attached to every call.

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please uncomment the line below and explain which changes are breaking.
-->
<!-- **This PR includes breaking changes to public APIs.** -->

<!--
Please uncomment the line below (and provide explanation) if the changes fix either (a) a security vulnerability, (b) a bug that caused incorrect or invalid data to be produced, or (c) a bug that causes a crash (even when the API contract is upheld). We use this to highlight fixes to issues that may affect users without their knowledge. For this reason, fixing bugs that cause errors don't count, since those are usually obvious.
-->
<!-- **This PR contains a "Critical Fix".** -->
* Closes: #33953
* Closes: #33839

aiguofer commented:
Unsure what that failure is... maybe the `ManagedChannel` shutdown is not being propagated to other channels? Or maybe it's not waiting long enough for the shutdown to complete (there's a hardcoded 5 sec wait there I believe)

lidavidm commented:
It'll be a while before I can review here but that test is known to be flaky

aiguofer commented:
Great, good to know! We're currently using a JDBC driver jar I built with these changes for development of our service so we're not currently blocked. Hoping these changes can make it into the next release so we can have users use a published jar once we get to beta. I really appreciate your help and time!

lidavidm commented on java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightClient.java:
```diff
@@ -353,10 +355,11 @@ public void onCompleted() {
 
     try {
-      final ClientCall<ArrowMessage, ArrowMessage> call = interceptedChannel.newCall(doExchangeDescriptor, callOptions);
+      final ClientCall<ArrowMessage, ArrowMessage> call = stub.getChannel().newCall(doExchangeDescriptor, callOptions);
```
Is it possible to factor out a helper that can take care of the wrapStub - getChannel - newCall dance?

aiguofer replied:
it'd be nice to add something like:

```
    public <RequestT, ResponseT> ClientCall<RequestT, ResponseT> newCall(MethodDescriptor<RequestT, ResponseT> descriptor) {
      return getChannel().newCall(descriptor, getCallOptions());
    }
```

to the stub so we could just do `stub.newCall(descriptor)`, but it looks like that class is auto-generated so I'm not sure where to add it.

lidavidm replied:
It can just be a private method, no?

aiguofer replied:
Oh I meant this would be great as a public method in `FlightServiceStub`. There's a lot of code in there that could use it, but I believe that is an auto-generated source file so I don't know how to change it.

I went ahead and created a private method in `FlightClient` to do the same.

lidavidm replied:
It's generated by gRPC, so it's not under our control

lidavidm reviewed (commented):
Can we add unit tests here?

aiguofer commented:
Sure I'll take a stab at writing some unit tests. I was thinking maybe something like `ConnectionTest` with a middleware that checks for the specific header values.

aiguofer commented on java/flight/flight-sql-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/CustomHeaderTest.java:
```diff
@@ -0,0 +1,118 @@
+
+      try {
+        // We don't care if the call fails, we just want to make sure the headers got passed
+        connection.getMetaData().getTableTypes();
```
It'd be great to have a sequence of calls that tests most grpc methods, but at this point I'm not familiar enough with Arrow SQL to write that.

Depending on what those calls might look like, it could be better to have multiple tests, but figured this simple test is at least a good starting point.

aiguofer replied:
Yeah makes sense, unfortunately I don't know the code well enough to do that. I know I encountered the issue by having properties in the JDBC connection that didn't make it to the server, and that's how I've been testing that the changes work.

Unfortunately I don't think I have the time right now to dig into it to write better unit tests... some help there would be great.

aiguofer commented:
@lidavidm Now that I'm a bit more familiar with the code base I was able to write some better unit tests. I moved the tests to `flight-core`, removed the use of rules, and I'm now testing each specific flight method instead of using the JDBC connection proxy.

aiguofer commented:
for the future... is there some command I can run to auto-format code to checkstyle? I haven't thought about checkstyles compliance in years, this should be automated.

lidavidm commented:
> for the future... is there some command I can run to auto-format code to checkstyle? I haven't thought about checkstyles compliance in years, this should be automated.

Unfortunately no. I would much prefer if we could switch to a formatter like google-java-format rather than checkstyle but that would be a somewhat disruptive refactor.

That said - feel free to raise an issue and get opinions from other Java contributors.

lidavidm reviewed (approved):
This is great, thank you!

lidavidm commented:
Actually, I filed https://github.com/apache/arrow/issues/34364 for checkstyle - feel free to chime in there
